src/python/pants/backend/helm/check/kubeconform/common.py

```diff
@@ -0,0 +1,152 @@
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.engine.env_vars import EnvironmentVars, EnvironmentVarsRequest
+from pants.engine.fs import CreateDigest, Digest, FileEntry
+from pants.engine.platform import Platform
+from pants.engine.process import FallibleProcessResult, Process, ProcessCacheScope
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+class KubeconformCheckRequest(CheckRequest, metaclass=ABCMeta):
+    tool_name = KubeconformSubsystem.name
+
+
+@dataclass(frozen=True)
+class RunKubeconformRequest:
+    field_set: KubeconformFieldSet
+    rendered_files: RenderedHelmFiles
+
+
+@dataclass(frozen=True)
+class KubeconformSetup:
+    binary: DownloadedExternalTool
+    env: FrozenDict[str, str]
+
+    @property
+    def cache_dir(self) -> str:
+        return "__kubeconform_cache"
+
+    @property
+    def append_only_caches(self) -> dict[str, str]:
+        return {"kubeconform": self.cache_dir}
+
+
+@rule
+async def setup_kube_conform(
+    helm: HelmSubsystem,
+    kubeconform: KubeconformSubsystem,
+    platform: Platform,
+) -> KubeconformSetup:
+    downloaded_tool, env = await MultiGet(
+        Get(DownloadedExternalTool, ExternalToolRequest, kubeconform.get_request(platform)),
+        Get(EnvironmentVars, EnvironmentVarsRequest(helm.extra_env_vars)),
+    )
+    return KubeconformSetup(downloaded_tool, env)
+
+
+@rule
+async def run_kubeconform(
+    request: RunKubeconformRequest, setup: KubeconformSetup, kubeconform: KubeconformSubsystem
+) -> CheckResult:
+    tool_relpath = "__kubeconform"
+    chart_relpath = "__chart"
+    immutable_input_digests = {
+        tool_relpath: setup.binary.digest,
+        chart_relpath: request.rendered_files.chart.snapshot.digest,
+    }
+
+    debug_requested = 0 < logger.getEffectiveLevel() <= LogLevel.DEBUG.level
+    ignore_sources = request.field_set.ignore_sources.value or ()
+
+    result = await Get(
+        FallibleProcessResult,
+        Process(
+            argv=[
+                os.path.join(tool_relpath, setup.binary.exe),
+                "-cache",
+                setup.cache_dir,
+                *(("-n", str(kubeconform.concurrency)) if kubeconform.concurrency else ()),
+                *(("-debug",) if debug_requested else ()),
+                *(("-summary",) if kubeconform.summary else ()),
+                *(("-verbose",) if kubeconform.verbose else ()),
+                *(("-strict",) if request.field_set.strict.value else ()),
+                *(
+                    ("-ignore-missing-schemas",)
+                    if request.field_set.ignore_missing_schemas.value
+                    else ()
+                ),
+                *chain.from_iterable(
+                    ("-ignore-filename-pattern", shlex.quote(pattern)) for pattern in ignore_sources
```

Comment:
Is `shlex.quote` necessary here? A `Process` usually does not imply a shell invocation and thus any quote marks will be passed into the invoked tool instead of unescaped by a shell. What does `setup.binary.exe` represent?

Reply:
it represents the sandboxed path to the `Kubeconform` binary. The _patterns_ being passed through `shlex` here are regexes so the reason I'm trying to escape them is to prevent being misinterpreted.

If the fact that `setup.binary.exe` points to an actual binary (and not to a shell or a script) already guarantees that these patterns won't be misread (or that being escaped could cause that misinterpretation) then the escaping is unnecessary.